pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1916,20 +1916,20 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
         if (messageIds == null || messageIds.isEmpty()) {
             return CompletableFuture.completedFuture(Collections.emptyList());
         }
-        List<MessageIdData> data = new ArrayList<>(messageIds.size());
-        List<CompletableFuture<Void>> futures = new ArrayList<>(messageIds.size());
-        messageIds.forEach(messageId -> {
+        List<CompletableFuture<MessageIdData>> futures = messageIds.stream().map(messageId -> {
             CompletableFuture<Boolean> future = processPossibleToDLQ(messageId);
-            futures.add(future.thenAccept(sendToDLQ -> {
+            return future.thenApply(sendToDLQ -> {
                 if (!sendToDLQ) {
-                    data.add(new MessageIdData()
+                    return new MessageIdData()
                             .setPartition(messageId.getPartitionIndex())
                             .setLedgerId(messageId.getLedgerId())
-                            .setEntryId(messageId.getEntryId()));
+                            .setEntryId(messageId.getEntryId());
                 }
-            }));
-        });
-        return FutureUtil.waitForAll(futures).thenCompose(v -> CompletableFuture.completedFuture(data));
+                return null;
+            });
+        }).collect(Collectors.toList());
+        return FutureUtil.waitForAll(futures).thenApply(v ->
+                futures.stream().map(CompletableFuture::join).filter(Objects::nonNull).collect(Collectors.toList()));
```

Comment:
`CompletableFuture::join` blocks.

Reply:
To implement a non-blocking reduction to a list, one would have to use `CompletableFuture.thenCombine` and reduce pair wise until there's a single result list. I didn't find a code example of this, but https://stackoverflow.com/a/65654939 is in that direction.
Most Stackoverflow examples suggest using CompletableFuture::join which isn't a non-blocking approach.

Reply:
I now found the library that I have used in the past: https://github.com/spotify/completable-futures#allaslist

It properly handles what is needed here.

Reply:
Actually Spotify Completable Futures library seems to be blocking and using CompletableFuture::join too, https://github.com/spotify/completable-futures/blob/b2175e5daf58d8882b5cd6e9eec62e4c459f9437/src/main/java/com/spotify/futures/CompletableFutures.java#L99 . That's not a non-blocking approach.

Reply:
Actually using CompletableFuture::join here is safe cause the `futures` are already **completed**, it return immediately instead of block waiting for result.

Reply:
And Spotify Completable Futures you mentioned is also using `CompletableFuture::join` in same way, it waits `allOf` complete first then call join to get result.

Reply:
> Actually using CompletableFuture::join here is safe cause the futures are already completed, it returns immediately instead of block waiting for result.

Yes that is correct.